src/metabase/search/core.clj

```diff
@@ -114,28 +117,41 @@
         (analytics/inc! :metabase-search/index-error)
         (throw e)))))
 
+(defonce ^:private reindex-pool
+  (ThreadPoolExecutor.
+   1 ; pool size
+   1 ; max pool size
+   0 TimeUnit/MILLISECONDS ; keepalive
+   (ArrayBlockingQueue. 1)
+   (ThreadPoolExecutor$DiscardPolicy.)))
+
 (defn reindex!
-  "Populate a new index, and make it active. Simultaneously updates the current index."
-  [& {:as opts}]
-  ;; If there are multiple indexes, return the peak inserted for each type. In practice, they should all be the same.
-  (when (supports-index?)
-    (try
-      (log/info "Reindexing searchable entities")
-      (let [timer (u/start-timer)
-            report (reduce (partial merge-with max)
-                           nil
-                           (for [e (search.engine/active-engines)]
-                             (search.engine/reindex! e opts)))
-            duration (u/since-ms timer)]
-        (analytics/inc! :metabase-search/index-reindex-ms duration)
-        (prometheus/observe! :metabase-search/index-reindex-duration-ms duration)
-        (doseq [[model cnt] report]
-          (analytics/inc! :metabase-search/index-reindexes {:model model} cnt))
-        (log/infof "Done reindexing in %.0fms %s" duration (sort-by (comp - val) report))
-        report)
-      (catch Exception e
-        (analytics/inc! :metabase-search/index-error)
-        (throw e)))))
+  "Populate a new index, and make it active. Simultaneously updates the current index.
+  Returns a future that will complete when the reindexing is done.
+  Respects `search.ingestion/*force-sync*` and waits for the future if it's true."
+  [& {:keys [async?] :or {async? true} :as opts}]
+  (let [logic (fn [] (when (supports-index?)
+                       (try
+                         (log/info "Reindexing searchable entities")
+                         (let [timer (u/start-timer)
+                               report (reduce (partial merge-with max)
+                                              nil
+                                              (for [e (search.engine/active-engines)]
+                                                (search.engine/reindex! e opts)))
+                               duration (u/since-ms timer)]
+                           (analytics/inc! :metabase-search/index-reindex-ms duration)
+                           (prometheus/observe! :metabase-search/index-reindex-duration-ms duration)
+                           (doseq [[model cnt] report]
+                             (analytics/inc! :metabase-search/index-reindexes {:model model} cnt))
+                           (log/infof "Done reindexing in %.0fms %s" duration (sort-by (comp - val) report))
+                           report)
+                         (catch Exception e
+                           (analytics/inc! :metabase-search/index-error)
+                           (throw e)))))]
+    (if (or search.ingestion/*force-sync* async?)
+      (let [result (logic)]
+        (future result))
```

Comment:
nit: it would be a bit nicer to just return a promise here, so we don't even run the constant on a background thread.